### PR TITLE
Reservations

### PR DESCRIPTION
**Pull request by Hamdy**

### CODE REVIEW

File: specs/raml/sdstor.raml [hunk 1]
```diff
@@ -152,6 +165,35 @@ types:
   post:
     body:
       type: NamespaceCreate
+
+  /stats:
+     securedBy: [ itsyouonline : { scopes: ["user:name"] } ]
+     post:
```
Comment (zaibon) [inline specs/raml/sdstor.raml]: Whats the use case of this ?
The store statistics is a read-only information for the user. It should not be allow to modify it.

File: store/ACLEntry.go [hunk 2]
```diff
@@ -16,3 +16,68 @@ func (s ACLEntry) Validate() error {
...
 }
+
+/*
+
+| Admin  | Read  | Write  | Delete  |
+|--------|-------|--------|---------|
+| 1 byte | 1 byte| 1 byte | 1 byte  |
+
+ */
+func(s ACLEntry) ToBytes() []byte{
```
Comment (zaibon) [inline store/ACLEntry.go]: bit mask would probably be faster

File: store/NamespaceCreate.go [hunk 3]
```diff
@@ -12,3 +13,126 @@ type NamespaceCreate struct {
...
+	return s.Save(db, config)
+
+}
+
+func (s NamespaceCreate) Exists(db *Badger, config *settings) (bool, error){
+	return db.Exists(s.Label)
+}
+
+func (s NamespaceCreate) Save(db *Badger, config *settings) error{
+	return db.Set(s.Label, s.ToBytes())
```
Comment (zaibon) [inline store/NamespaceCreate.go]: Shouldn't we also have a prefix for Namespace ?
So we can list them really fast ?
Cause now you have to walk over the all database to list the namespaces

File: store/main.go [hunk 4]
```diff
@@ -110,6 +131,28 @@ func main() {
 			Value:       100,
 			Destination: &settings.Iterator.PreFetchSize,
 		},
+
+		cli.StringFlag{
```
Comment (zaibon) [inline store/main.go]: I don't think we should make these prefix configurable.
This is internal business logic.

File: store/namespaces_api_ListReservations.go [hunk 5]
```diff
@@ -2,9 +2,101 @@ package main
...
+	opt.PrefetchSize = api.config.Iterator.PreFetchSize
+
+	it := api.db.store.NewIterator(opt)
+	defer it.Close()
+
+	startingIndex := (page-1)*per_page + 1
+	counter := 0 // Number of objects encountered
+	resultsCount := per_page
+
+	for it.Rewind(); it.Valid(); it.Next() {
```
Comment (zaibon) [inline store/namespaces_api_ListReservations.go]: Use `it.Seek(prefix)` and stop iteration when the prefix is not present anymore.

Comment (zaibon) [inline store/namespaces_api_ListReservations.go]: See this commit: they added a new method on the iterator object to do exactly this:
https://github.com/dgraph-io/badger/commit/eaad7ebcf7437492bdf65ce674976a0fd531e71e

File: specs/raml/sdstor.raml (hunk 1 above)
Comment (Hamdy) [inline specs/raml/sdstor.raml]: Mistake, this should have been a separate section
so basically there's a store stats which holds the available total size that store can handle and we can change that limit in runtime through a rest call.
It holds also total amount of space used for all reservations. 
this is efficient and enables us to determine whether a new reservation is applicable or not without going through all reservation records in db and get the total amount of space used for valid ones in runtime when creating new reservations.

File: store/ACLEntry.go (hunk 2 above)
Comment (Hamdy) [inline store/ACLEntry.go]: can u explain how I can do bit masking in this case?

File: store/main.go (hunk 4 above)
Comment (Hamdy) [inline store/main.go]: OK

File: store/NamespaceCreate.go (hunk 3 above)
Comment (Hamdy) [inline store/NamespaceCreate.go]: Listing namespaces is not a frequent operation. it's meant for admin of the 0-stor not the normal users.

File: store/namespaces_api_ListReservations.go (hunk 5 above)
Comment (Hamdy) [inline store/namespaces_api_ListReservations.go]: OK

File: store/NamespaceCreate.go (hunk 3 above)
Comment (zaibon) [inline store/NamespaceCreate.go]: Doesn't matter if it's called a lot or not, walking the full database is bad

File: specs/raml/sdstor.raml (hunk 1 above)
Comment (zaibon) [inline specs/raml/sdstor.raml]: >  and we can change that limit in runtime through a rest call.

This is not a requirement at the moment, and + if we do that, then it needs to be protected with a jwt that confirm we are the admin of the store, so ` [ itsyouonline : { scopes: ["user:name"] } ]` is wrong.

>this is efficient and enables us to determine whether a new reservation is applicable or not without going through all reservation records in db and get the total amount of space used for valid ones in runtime when creating new reservations.

This is great but this should be some internal knowledge that the store need for himself. I don't get why it needs to be exposed.

File: store/ACLEntry.go (hunk 2 above)
Comment (zaibon) [inline store/ACLEntry.go]: https://play.golang.org/p/Dm3X-akr31

**Comment by zaibon** [comment]
Merging this cause quite some logic is part of this PR and it would be hard to split it now.
but some refactoring on the code is needed and going to be provided in next PR

**Pull request merged**